cache/disk/load.go

```diff
@@ -0,0 +1,364 @@
+package disk
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"sort"
+	"sync"
+
+	"github.com/djherbis/atime"
+
+	"github.com/buchgr/bazel-remote/cache"
+
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+)
+
+type scanDir struct {
+	name    string
+	dest    string
+	version int
+	kind    cache.EntryKind
+}
+
+// Return a list of importItems sorted by atime, and a boolean that is
+// true if the caller should migrate items, in reverse LRU order.
+func (c *DiskCache) findCacheItems() ([]importItem, bool, error) {
+	files := []importItem{}
+
+	var mu sync.Mutex // Protects the migrate variable below:
+	migrate := false
+
+	// Workers submit discovered files here.
+	filesChan := make(chan []importItem)
+
+	// Workers receive a dir to scan here.
+	workChan := make(chan scanDir)
+
+	// Workers can report errors here:
+	errChan := make(chan error)
+
+	numWorkers := runtime.NumCPU() // TODO: consider tweaking this.
+
+	hashKeyRegex := regexp.MustCompile("^[a-f0-9]{64}$")
+
+	// Spawn some worker goroutines to read the cache concurrently.
+	var wg sync.WaitGroup
+	for i := 0; i < numWorkers; i++ {
+		wg.Add(1)
+		go func(i int) {
+			needMigration := false
+			defer func() {
+				if needMigration {
+					mu.Lock()
+					migrate = true
+					mu.Unlock()
+				}
+				wg.Done()
+			}()
+
+			for scanDir := range workChan {
+				_, err := os.Stat(scanDir.name)
+				if os.IsNotExist(err) {
+					continue
+				} else if err != nil {
+					errChan <- err
+					return
+				}
+
+				listing, err := ioutil.ReadDir(scanDir.name)
+				if err != nil {
+					errChan <- err
+					return
+				}
+
+				addChecksum := scanDir.version < 2 && (scanDir.kind != cache.CAS)
+
+				toSend := make([]importItem, 0, len(listing))
+				for e := range listing {
+					if listing[e].IsDir() {
+						continue
+					}
+
+					if !hashKeyRegex.MatchString(listing[e].Name()) {
+						log.Println("Unexpected file in cache:",
+							filepath.Join(scanDir.name, listing[e].Name()))
+						continue
+					}
+
+					basename := listing[e].Name()
+					entry := importItem{
+						name:        filepath.Join(scanDir.name, basename),
+						info:        listing[e],
+						addChecksum: addChecksum,
+					}
+
+					if scanDir.version < 2 {
```

Comment:
Why is it necessary to mix the migration detection with the loading logic? Shouldn't detecting whether a migration is necessary be as simple as checking whether a folder V-1 exists and is not empty?

```
if hasItems("ac/)) {
  migrate();
  rmdir("ac/")
}

load_items()
```

Reply:
Good idea.

Reply:
I looked through this code again, and remembered why I mixed the logic:
1) So we end up with an accurate order for the LRU index (migrating old style files first would always place them at the start of the index even if they're older than new style files).
2) Building the LRU index is slower when done sequentially than in parallel, and I would prefer cache upgrades to be fast (less chance that the admin might think something went wrong and ctrl-c in the middle of an upgrade).